src/core/border_router/routing_manager.hpp

```diff
@@ -0,0 +1,259 @@
+#define ROUTING_MANAGER_HPP_
+
+#include "openthread-core-config.h"
+
+#if OPENTHREAD_CONFIG_BORDER_ROUTING_ENABLE
+
+#include <openthread/error.h>
+
+#include "common/locator.hpp"
+
+#include "border_router/router_advertisement.hpp"
+#include "common/notifier.hpp"
+#include "common/timer.hpp"
+#include "net/ip6.hpp"
+
+namespace ot {
+
+/**
+ * This class implements bi-directional routing between Thread and
+ * Infrastructure networks.
+ *
+ * The routing manager works on both Thread interface and infrastructure
+ * interface. All ICMPv6 messages are sent/recv on the infrastructure
+ * interface.
+ *
+ */
+class RoutingManager : public InstanceLocator
+{
+    friend class ot::Notifier;
+
+public:
+    /**
+     * This constructor initializes the routing manager.
+     *
+     * @param[in]  aInstance  A OpenThread instance.
+     *
+     */
+    explicit RoutingManager(Instance &aInstance);
+
+    /**
+     * This method initializes the routing manager on given infrastructure interface.
+     *
+     * @param[in]  aInfraIfIndex  An infrastructure network interface index.
+     * @param[in]  aInfraIfName   An infrastructure network interface index.
+     *
+     * @retval  OT_ERROR_NONE          Successfully started the routing manager.
+     * @retval  OT_ERROR_INVALID_ARGS  The index or name of the infra interface is not valid.
+     * @retval  OT_ERROR_ALREADY       The routing manager has already been initialized.
+     *
+     */
+    otError Init(uint32_t aInfraIfIndex, const char *aInfraIfName);
+
+    /**
+     * This method receives an ICMPv6 message on the infrastructure interface.
+     *
+     * Malformed or undesired messages are dropped silently.
+     *
+     * @param[in]  aInfraIfIndex  The infrastructure interface index.
+     * @param[in]  aSrcAddress    The source address this message is sent from.
+     * @param[in]  aBuffer        THe ICMPv6 message buffer.
+     * @param[in]  aLength        The length of the ICMPv6 message buffer.
+     *
+     */
+    void RecvIcmp6Message(uint32_t            aInfraIfIndex,
+                          const Ip6::Address &aSrcAddress,
+                          const uint8_t *     aBuffer,
+                          uint16_t            aBufferLength);
+
+private:
+    static constexpr uint8_t kMaxInfraIfNameLength = 16;
+
+    static constexpr uint16_t kMaxRouterAdvMessageLength = 256;
+
+    static constexpr uint32_t kDefaultOmrPrefixLifetime    = 1800;
+    static constexpr uint32_t kDefaultOnLinkPrefixLifetime = 1800;
+
+    static constexpr uint8_t kOmrPrefixLength    = OT_IP6_PREFIX_BITSIZE; ///< The length of an OMR prefix. In bits.
+    static constexpr uint8_t kOnLinkPrefixLength = OT_IP6_PREFIX_BITSIZE; ///< The length of an On-link prefix. In bits.
+
+    static constexpr uint32_t kMinRtrAdvInterval = 30;   ///< Minimum Router Advertisement Interval. In Seconds.
+    static constexpr uint32_t kMaxRtrAdvInterval = 1800; ///< Maximum Router Advertisement Interval. In Seconds.
```

Comment:
```suggestion
    static constexpr uint32_t kMaxRtrAdvInterval = 1800; ///< Maximum Router Advertisement Interval. In Seconds.
    static_assert(kMaxRtrAdvInterval > kMinRtrAdvInterval);
```

just in case.